core/genallocs/genesis_alloc.go

```diff
@@ -0,0 +1,128 @@
+	VestSchedule    int                 `json:"Vest Schedule" gencodec:"required"`
+	Address         common.Address      `json:"Address" gencodec:"required"`
+	TotalBalance    uint64              `json:"Amount" gencodec:"required"`
+	BalanceSchedule map[uint64]*big.Int // Map of blockNumber->balanceUnlocked (at that block).
+	privateKey      []byte              `json:"secretKey,omitempty"`
+}
+
+type vestingSchedule struct {
+	vestDuration  uint64  // Total vesting duration in years. The first year cliff is not part of vesting.
+	tgePercentage float32 // One-time percentage unlocked at TGE.
+	unlockStart   int     // First month that regular unlocks start.
+}
+
+var vestingSchedules = [3]vestingSchedule{
+	{
+		// schedule1
+		// 	Vesting duration: 5 years
+		// 	TGE: 30%
+		vestDuration:  5,
+		tgePercentage: 0.3,
+	},
+
+	{
+		// schedule2
+		// 	Vesting duration: 3 years
+		// 	TGE: 25%
+		vestDuration:  3,
+		tgePercentage: 0.25,
+	},
+
+	{
+		// schedule3
+		// 	Vesting duration: 3 years
+		// 	TGE: 0%
+		vestDuration:  3,
+		tgePercentage: 0,
+	},
+}
+
+// Will return all the GenesisAccounts with their calculated vesting schedules
+func AllocateGenesisAccounts(filename string) ([]GenesisAccount, error) {
+	// Read from allocs file.
+	allocs, err := readGenesisAllocs(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate vesting schedules for each account.
+	for i := range allocs {
+		allocs[i].calculateLockedBalances()
+	}
+	return allocs, nil
+}
+
+// Parses the allocs JSON and populates the basic vesting info.
+func readGenesisAllocs(filename string) ([]GenesisAccount, error) {
+	// Open the JSON file
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %w", err)
+	}
+	defer file.Close()
+
+	return decodeGenesisAllocs(file)
+}
+
+func decodeGenesisAllocs(r io.Reader) ([]GenesisAccount, error) {
+	// Decode into slice
+	var accounts []GenesisAccount
+	decoder := json.NewDecoder(r)
+	if err := decoder.Decode(&accounts); err != nil {
+		return nil, fmt.Errorf("failed to decode JSON: %w", err)
+	}
+
+	return accounts, nil
+}
+
+// Will calculate the unlock heights according to the pre-defined vesting schedule.
+func (account *GenesisAccount) calculateLockedBalances() {
+	account.BalanceSchedule = make(map[uint64]*big.Int)
+
+	vestingSchedule := vestingSchedules[account.VestSchedule]
+	balanceInt := new(big.Int).SetUint64(account.TotalBalance)
+
+	// Calculate total unlock at tge.
+	tgePercentage := big.NewInt(int64(vestingSchedule.tgePercentage * 100)) // Multiply by 100 to remove need for floats.
+	tgeAmount := new(big.Int).Mul(balanceInt, tgePercentage)
+	tgeAmount.Div(tgeAmount, common.Big100) // Divide back by 100.
+	account.BalanceSchedule[0] = tgeAmount
+
+	// Calculate number of unlocks.
+	numUnlocks := uint64((vestingSchedule.vestDuration) * 12) // Total months that unlock.
+
+	// Calculate amount per unlock.
+	quaiPerUnlock := new(big.Int).Sub(balanceInt, tgeAmount)
```

Comment:
I don't think it changes the actual math, but to be clear, this is the number of Wei (quei, qwei, whatever its called) to unlock, not Quai.

As long as the genesis alloc amounts are denominated in Wei, then we I think all the math remains the same, but could you confirm?